tracer/src/Datadog.Trace/Logging/DirectSubmission/Sink/PeriodicBatching/BatchedConnectionStatus.cs

```diff
@@ -0,0 +1,107 @@
+﻿// <copyright file="BatchedConnectionStatus.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+// Based on https://github.com/serilog/serilog-sinks-periodicbatching/blob/66a74768196758200bff67077167cde3a7e346d5/src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/BatchedConnectionStatus.cs
+// Copyright 2013-2020 Serilog Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+
+namespace Datadog.Trace.Logging.DirectSubmission.Sink.PeriodicBatching
+{
+    /// <summary>
+    /// Manages reconnection period and transient fault response for <see cref="BatchingSink"/>.
+    /// During normal operation an object of this type will simply echo the configured batch transmission
+    /// period. When availability fluctuates, the class tracks the number of failed attempts, each time
+    /// increasing the interval before reconnection is attempted (up to a set maximum) and at predefined
+    /// points indicating that either the current batch, or entire waiting queue, should be dropped. This
+    /// Serves two purposes - first, a loaded receiver may need a temporary reduction in traffic while coming
+    /// back online. Second, the sender needs to account for both bad batches (the first fault response) and
+    /// also overproduction (the second, queue-dropping response). In combination these should provide a
+    /// reasonable delivery effort but ultimately protect the sender from memory exhaustion.
+    /// </summary>
+    /// <remarks>
+    /// Currently used only by <see cref="BatchingSink"/>, but may
+    /// provide the basis for a "smart" exponential backoff timer. There are other factors to consider
+    /// including the desire to send batches "when full" rather than continuing to buffer, and so-on.
+    /// </remarks>
+    internal class BatchedConnectionStatus
+    {
```

Comment:
I don't think I am comfortable with the scale of the exponential backoff here.
We could be piling up for several minutes it seems like.
Minimally I think these retries should be capped to something like 4, and at best case they should be configurable.

Reply:
Yeah, 10 minutes does seem like a long time. That said, if there's no connection to be made, what's the best option? We have a size limit on the queue size, so we're not going to blow up indefinitely, and arguably those _first_ logs when things started to go wrong might be more important than subsequent ones, so maybe better to hang onto those and drop new logs? Not sure what's best tbh, though agree configurable is a good way we can blame it on the user when it's wrong 😉

Reply:
Actually, looking through this again, there's significant overlap with the `BatchedConnectionStatus` and the `LogsApi`. **Both** are doing retries/exponential backoff, which we definitely don't want. That said, we're never throwing in `DatadogSink.EmitBatch()`, so none of the retry code in `BatchedConnectionStatus` will ever run

I'm inclined to do the following:
* Drop `BatchedConnectionStatus` entirely
* The `LogsApi` has better knowledge about when a batch should be dropped etc, so I think `SendLogsAsync` should return a result with what to do (`Success`/`DropBatch`/`DropQueue`). That can bubble up via `EmitBatch()` so we can remove `BatchedConnectionStatus`. 

Alternatively, rip the whole thing out and do our own version 😬 The Serilog API _is_ cleaner in some respects, as it separates the logic for when to retry etc from the implementors, but it's using exceptions for flow control, and we're handling that ourselves anyway.

What do you think?
1. Rip out `BatchedConnectionStatus` only?
2. Tear it all down?
3. Do 1. or 2. in a follow-up PR

Reply:
Yeah, I can tear it out as part of this PR. I guess it just adds confusion for other people looking at it now anyway 👍

Reply:
Reworked the vast majority of the batching implementation in https://github.com/DataDog/dd-trace-dotnet/pull/1945/commits/2dfc44ef19f69044cb8065ee395557ccd19fc6b3.

* Dropped `BatchedConnectionStatus`
* Dropped `PeriodicTimer`
* Added `CircuitBreaker`
* Removed some of the complexity around starting/stopping the sink with a tcs. We could optimise it more, but seems sufficient for now to me.